### ISSUE TITLE
Differences from meekrodb

### ISSUE THREAD

**Issue by aolko** [2blane/IronSQL#1]
And the differences are...?

**Comment by aolko** [comment]
Oh come on, you've just "fixed" the composer and that's it? Why not fix assoc array output?

**Comment by aolko** [comment]
@2blane hello, any response?

**Comment by 2blane** [comment]
MeekroDB looks like nobody is taking care of that project so I forked because I needed the ability to switch databases on the same MySQL Database host and that wasn't supported. Also, I needed an insertIgnoreOrUpdate command which wasn't supported.

**Comment by aolko** [comment]
well, you see, so far it feels just as dead as meekrodb @2blane 

- will tough spots where injections might be possible get improved?
- will it get it's own docs?
- will it have assoc arrays figured out? (Why? Template engines, they use array data) (no wrapping in `$array[0]` deals for nested arrays) (currently i have to use separate tricks for `select one` and `select many` cases)

**Comment by 2blane** [comment]
1 - might be possible or actually happen? If there are spots where injections happen I’d like to know.
2 - this is a good idea I’ll work on that
3 - can you give a more concrete example of this and I’ll see what I can do?

**Comment by aolko** [comment]
1. Well people will argue the usual - "PDO is safer", "PDO is better", etc.
3. This kind of tricks
```
public static function getDBdata($query, $type="multiple"){
            $raw = DB::query($query);
            $refined = array();
            if ($type=="multiple"){
                foreach ($raw as $row)
                {
                    $refined[] = array_map("stripslashes",$row);
                }
                return $refined;
            }
            elseif ($type=="single"){
                $raw = DB::queryRaw($query);
                $refined = $raw->fetch_assoc();
                return $refined;
            }
        }
```
gets really annoying over time, it'd be great if the library could figure that out by itself, while you are querying it via `DB::query()`

Also, is there a possibility of introducing [Relations](https://www.notorm.com/) to ironsql?

**Comment by 2blane** [comment]
I’m still not sure what you’re asking? There is a method to query for one row vs. multiple rows? queryFirstRow() is what it’s called I believe? Or are you saying if there is only one row of data returned from the query it should not be returned in an array and should just be unwrapped?

**Comment by aolko** [comment]
> I’m still not sure what you’re asking? There is a method to query for one row vs. multiple rows? queryFirstRow() is what it’s called I believe? Or are you saying if there is only one row of data returned from the query it should not be returned in an array and should just be unwrapped?

I'm saying that ordinary `DB::query()` wraps result in an array, if result is already a nested array - that hurts, i don't need that. Therefore i'm relying on tricks. To not rely on tricks make the lib figure it out by itself. And yes, i'm always selecting all the rows.

**Comment by 2blane** [comment]
Can you provide your query and an example response? I don’t quite understand how your query result is an array unless it is JSON or some similar format?

**Comment by aolko** [comment]
let's say
```
DB::QueryRaw("SELECT * FROM books");
```
```
$books = [
["name"=>"Book 1"],["name"=>"Book 2"]
];
```
```
{% for book in books %}{{ book.name }}<br/>{% endfor %}
```
```
Book 1
Book 2
```

**Comment by 2blane** [comment]
Ok, so with your example what is the issue? That is exactly how it is supposed to work? You have an array of results named $books. And you loop through each one and display the name? Or do you want the result to be an array of bookNames like [“Book 1”, “Book 2”]?

**Comment by aolko** [comment]
> Ok, so with your example what is the issue? That is exactly how it is supposed to work? You have an array of results named $books. And you loop through each one and display the name? Or do you want the result to be an array of bookNames like [“Book Name1”, “Book Name2”]?

i want a nested assoc array as output, so it can be accessed by twig w/o any problems and can be iterated and called.
`DB::query()` mangles it by wrapping it in array __on top of that__, so we have an `array[0][n]["name"]`

tl;dr: if you really want to check that, try shoving `DB::query()` output into a twig template with a for loop

**Comment by 2blane** [comment]
Got it, so I’ve never experienced that problem and I don’t use Twig. When you run DB::query() it is expected that you get back an array of associated arrays. For instance, if you ran DB::query() like the following:
`
$results = DB::query(“SELECT * FROM BOOKS”);
`
Results would then be:
[
    [“name” => “Book1”],
    [“name” => “Book2”],
    ...
]

For queryRaw(), you get back a mysqli_result. So, you then have to call fetch_assoc just like you are doing in your examples above.

I would implore you to print out the results of DB::query() and see if they match the format that I’ve shown here. If they do not and the array[0][n][“name”] problem exists then it is a problem with your query and you are somehow returning all of your data as one row. If that is the case, use something like Sequel Pro to debug your query. If the problem happens later (with Twig), then it is something you are doing between the result be returned by the database and your interaction with Twig.

**Comment by 2blane** [comment]
Looked up Twig real quick. I think I know what your problem is.
You used:
`{% for book in books %}{{ book.name }}<br/>{% endfor %}`
So with PHP, when you have an associative array like:
[
[“name” => “Book1”],
[“name” => “Book2”],
...
]
What really happens is PHP keeps an index to each array value like:
[
0 => [“name” => “Book1”],
1 => [“name” => “Book2”],
...
]
So when you loop over books for instance you could be getting 0, 1, 2, ... So, what you can do is change your for loop to give you back both the key and value like so:
`{% for key, book in books %}{{ book.name }}<br/>{% endfor %}`

**Comment by aolko** [comment]
> So when you loop over books for instance you could be getting 0, 1, 2, ... So, what you can do is change your for loop to give you back both the key and value like so:
> `{% for key, book in books %}{{ book.name }}<br/>{% endfor %}`

true, yet preferable way is to iterate as usual, while modifying the output.
`{% for book in books %}{{ book.name }}<br/>{% endfor %}`

so, essentially, either flatten the arrays in the array or remove the wrapping array

**Comment by 2blane** [comment]
Yah, this isn't a problem with the library but with your implementation. It's hard to point to the exact line of code since you haven't provided the complete picture. Also, your function getDBData() doesn't do anything but call stripslashes on the data if $type=="multiple". You also shouldn't be using queryRaw() and then putting all the results in memory. queryRaw() is for when there are too many results and you want to process them one at a time because it wouldn't be feasible to load all of them in memory.
